apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,218 @@
+	//
+	// GitHub team slugs are lower case alphanumeric characters and may contain dashes (e.g. "kube-admins").
+	//
+	// Either way, group names as presented to Kubernetes will always be prefixed by the GitHub organization
+	// name followed by a forward slash (e.g. "my-org/my-team"). GitHub organization login names can only contain
+	// alphanumeric characters or single hyphens, so the first forward slash `/` will be the separator between
+	// the organization login name and the team name or slug.
+	//
+	// If desired, an admin could configure identity transformation expressions on a FederationDomain to
+	// further customize how these group names are presented to Kubernetes.
+	//
+	// +kubebuilder:default=slug
+	// +kubebuilder:validation:Enum=name;slug
+	Groups GitHubGroupNamesSpec `json:"groups"`
+}
+
+// GitHubClientSpec contains information about the GitHub client that this identity provider will use
+// for web-based login flows.
+type GitHubClientSpec struct {
+	// SecretName contains the name of a namespace-local Secret object that provides the clientID and
+	// clientSecret for an GitHub App or GitHub OAuth2 client.
+	//
+	// This secret must be of type "secrets.pinniped.dev/github-client" with keys "clientID" and "clientSecret".
+	SecretName string `json:"secretName"`
+}
+
+// GitHubIdentityProviderSpec is the spec for configuring an GitHub identity provider.
+type GitHubIdentityProviderSpec struct {
+	// GitHubApi allows configuration for GitHub Enterprise Server
+	//
+	// +optional
+	GitHubApi GitHubApiConfig `json:"github_api,omitempty"`
+
+	// Claims allows customization of the username and groups claims
+	//
+	// +optional
+	Claims GitHubClaims `json:"claims,omitempty"`
+
+	// AllowedOrganizations, when specified, indicates that only users with membership in at least one of the listed
+	// GitHub organizations may log in. In addition, the group membership presented to Kubernetes
+	// will only include teams within the listed GitHub organizations. Additional login rules or group filtering
+	// can optionally be provided as policy expression on any FederationDomain that includes this IDP.
+	//
+	// The configured GitHub App or GitHub OAuth App must be allowed to see membership in the listed organizations,
+	// otherwise Pinniped will not be aware that the user belongs to the listed organization or any teams
+	// within that organization.
+	//
+	// If no organizations are listed, you must set allowAllUsersToLogin to "Allow".
+	//
+	// +optional
+	AllowedOrganizations []string `json:"allowedOrganizations"`
+
+	// AllowAllUsersToLogin must be set to "Allow" if allowedOrganizations is empty.
+	//
+	// This field only exists to ensure that Pinniped administrators are aware that an empty list of
+	// allowedOrganizations means all GitHub users are allowed to log in.
+	//
+	// +kubebuilder:default=Deny
+	// +kubebuilder:validation:Enum=Allow;Deny
+	AllowAllUsersToLogin GitHubAllowAllUsersToLoginSpec `json:"allowAllUsersToLogin"`
```

Comment:
Is the naming sort of awkward that you can set a field called "AllowSomething" to have the value "Deny"? The field's name says it is about allowing something, not denying something.

Maybe name it something like `AllowedUsers`, to be symmetrical with `AllowedOrganizations`? Possible values would be something like:
- `OnlyUsersBelongingToAllowedOrganizations` (the default), which also makes it clear that it relates back to the `AllowedOrganizations` field
- `AllUsers`

Reply:
It's definitely a little awkward.

How about `RestrictGitHubUsersToAllowedOrganizations` with `Yes` as the default and `No` as the other option?

Reply:
I find it confusing that those names have substrings `AllowedOrganizations` and `OrganizationsAllowed` which makes them sound as if they are the same thing at first glance.

Reply:
Hmmm that makes sense. Let's chat tomorrow and capture the decision here.

Reply:
How about `AllowedOrganizationsPolicy`? That name ties it back to the `AllowedOrganizations` field.

We want the non-default value to emphasize the effect that it allows all GitHub users to authenticate regardless of org membership, even if they belong to no orgs. So how about enum values like `OnlyUsersBelongingToSpecifiedOrgs` and `AllGitHubUsers`.

So the settings in practice would look like:

```yaml
spec:
  allowedOrganizations: [vmware-tanzu, broadcom]
  allowedOrganizationsPolicy: OnlyUsersBelongingToSpecifiedOrgs
```

or

```yaml
spec:
  allowedOrganizations: []
  allowedOrganizationsPolicy: AllGitHubUsers
```

and because of the default values of the two settings, these would also be valid:

spec:
  allowedOrganizations: [vmware-tanzu, broadcom]
```

and

spec:
  allowedOrganizationsPolicy: AllGitHubUsers
```

The only configurations that allow all GitHub users must explicitly say `AllGitHubUsers` because it is the non-default value. Kinda verbose but hopefully very clear.

Reply:
I like it

Reply:
Let's do this instead:

```yaml
---
spec:
  client:
    secretName: foo
  allowAuthentication:
    organizations:
      policy: OnlyUsersFromAllowedOrganizations
      allowed:
      - foo
      - vmware-tanzu
      - Broadcom
    personalAccessTokens:
      fineGrained: true
---
spec:
  client:
    secretName: foo
  allowAuthentication:
    organizations:
      policy: AllGitHubUsers
    personalAccessTokens:
      classic: true
```